File: six/Bookmarks/Bookmark.swift
import Foundation
import SQLiteData

/// A saved page. The row is the record; the readable copy is a Markdown file in the profile's
/// `Bookmarks` folder (`fileName`), and the searchable form is `bookmark_chunks` + the vectors.
@Table("bookmarks")
nonisolated struct Bookmark: Identifiable, Sendable, Hashable {
    let id: UUID
    var profileID: UUID
    var url: URL
    var title = ""
    /// The page's own description, or the first paragraph.
    var excerpt = ""
    var siteName = ""
    var imageURL: URL?
    /// Name of the Markdown file inside the profile's bookmarks folder.
    var fileName = ""
    /// BCP-47 tag as the page declared it (or as detected), e.g. `ru`, `en`.
    var language = ""
    /// Length of the readable text — a hint of how much was saved.
    var characterCount = 0
    var createdAt: Date
    /// When the chunks were embedded; nil until the index has caught up (or `indexError` says why not).
    var indexedAt: Date?
    /// The embedder and index version the vectors were made with; a change means re-embedding.
    var embeddingModel = ""
    var indexError: String?
    /// When the page was last re-read from the site (nil: never since it was saved).
    var refreshedAt: Date?
    /// SHA-256 of the readable text, so a refresh that finds the same page changes nothing.
    var contentHash = ""
    /// Why the last refresh didn't happen — the site was down, the page has no text now.
    var refreshError: String?

    var lastReadAt: Date { refreshedAt ?? createdAt }

    var displayTitle: String { title.isEmpty ? url.absoluteString : title }
    var displayDetail: String { siteName.isEmpty ? (url.host() ?? url.absoluteString) : siteName }
}

/// One passage of a bookmark's text, in reading order. Chunk 0 is the title and excerpt, so a
/// search for what a page is about finds it even when the body is long.
@Table("bookmark_chunks")
nonisolated struct BookmarkChunk: Identifiable, Sendable {
    let id: UUID
    var bookmarkID: UUID
    var ord: Int
    var text: String
}

/// Which bookmarks a search or an assistant sees: the current profile's, or everyone's.
nonisolated enum BookmarkScope: String, CaseIterable, Identifiable, Sendable {
    case profile
    case all

    var id: String { rawValue }

    var title: String {
        #if os(Linux)
        // `String(localized:)` and the strings catalog behind it are Apple Foundation's; a GTK front
        // localises through gettext, so these are the keys and it translates them itself.
        switch self {
        case .profile: "This Profile"
        case .all: "All Profiles"
        }
        #else
        switch self {
        case .profile: String(localized: "This Profile")
        case .all: String(localized: "All Profiles")
        }
        #endif
    }
}

/// Which model the vectors are made with.
///
/// The index is bound to it: `Embedding.model` travels with every vector, `Bookmark.embeddingModel`
/// with every row, and `BookmarkStore.indexSignature` compares the two — so this is never a
/// conversion. Changing it re-embeds the library, and the two spaces are never compared.
///
/// The ladder is short on purpose. Both are E5 over the same 250 k vocabulary and the same hundred
/// languages; what the bigger one buys is ranking across languages, and what it costs — download,
/// memory, time — goes up with the hidden size, all three at once.
nonisolated enum EmbeddingModelChoice: String, CaseIterable, Identifiable, Sendable {
    /// `intfloat/multilingual-e5-small`: 118 M parameters, 384 dimensions, ~465 MB on disk.
    case small
    /// `intfloat/multilingual-e5-base`: 278 M parameters, 768 dimensions, ~1.1 GB on disk.
    case base

    var id: String { rawValue }

    /// The Hugging Face repository the weights come from.
    var repository: String {
        switch self {
        case .small: "intfloat/multilingual-e5-small"
        case .base: "intfloat/multilingual-e5-base"
        }
    }

    /// What a vector is stamped with, and what names the `vec0` table it goes into.
    var modelID: String {
        switch self {
        case .small: "multilingual-e5-small"
        case .base: "multilingual-e5-base"
        }
    }

    var dimension: Int {
        switch self {
        case .small: 384
        case .base: 768
        }
    }

    /// What this machine is offered before anybody chooses.
    ///
    /// Memory, and nothing else. The weights are held for as long as six runs, in memory the GPU and
    /// every WebKit process share: `small` holds ~235 MB of it and `base` about twice that, and on a
    /// Mac that has 8 GB — where macOS already sits in its `.warning` pressure band — the second one
    /// is paid for by the pages, which is the wrong thing to pay with. 16 GB is where that stops
    /// being true. Cores are deliberately not part of this: the ranking is what the bigger model
    /// buys, and a slower machine wants it no less.
    static var recommended: EmbeddingModelChoice {
        ProcessInfo.processInfo.physicalMemory >= 16 * 1024 * 1024 * 1024 ? .base : .small
    }

    /// The name in the picker. The size is in it because the size is the decision.
    var title: String {
        #if os(Linux)
        switch self {
        case .small: "Compact — 465 MB"
        case .base: "Larger — 1.1 GB"
        }
        #else
        switch self {
        case .small: String(localized: "Compact — 465 MB")
        case .base: String(localized: "Larger — 1.1 GB")
        }
        #endif
    }
}

/// A search result: the bookmark, how well it matched, and the passage that matched.
nonisolated struct BookmarkHit: Identifiable, Sendable {
    var bookmark: Bookmark
    /// 0…1, higher is better. Vector hits are `1 - cosine distance / 2`; text matches are fixed.
    var score: Double
    var snippet: String

    var id: Bookmark.ID { bookmark.id }
}

// MARK: - Settings

/// The setting lives in the settings table; the knowledge of what its string means lives here,
/// beside the type it means it as. `SettingsStore` itself keeps only keys and strings.
extension SettingsStore {
    /// What the assistant and the agents search: this profile's bookmarks, or every profile's.
    var bookmarkScope: BookmarkScope {
        get { BookmarkScope(rawValue: self[.bookmarkScope] ?? "") ?? .profile }
        set { self[.bookmarkScope] = newValue.rawValue }
    }

    /// The model the bookmarks are embedded with — nil until six has decided.
    ///
    /// Nil rather than a default on purpose: the decision is made once, at the first launch that
    /// asks, from `EmbeddingModelChoice.recommended` *and* from whether an index already exists, and
    /// then written down. An update is not allowed to re-embed somebody's library behind their back
    /// because the recommendation moved.
    var embeddingModel: EmbeddingModelChoice? {
        get { self[.embeddingModel].flatMap(EmbeddingModelChoice.init(rawValue:)) }
        set { self[.embeddingModel] = newValue?.rawValue }
    }
}

File: android/core/src/main/kotlin/org/deffun/six/core/Bookmark.kt
package org.deffun.six.core

import java.security.MessageDigest
import java.time.Instant
import java.time.format.DateTimeFormatter
import java.util.UUID

/**
 * A saved page. The row is the record; the readable copy is a Markdown file in the profile's
 * `Bookmarks` folder, and the searchable form is `bookmark_chunks`.
 *
 * Every column the Mac's table has, including the four this platform does not fill in yet —
 * `indexedAt`, `embeddingModel`, `indexError` and the refresh triple. They are read and written back
 * unchanged rather than dropped: a bookmark saved on the Mac, opened here and edited would otherwise
 * come back having forgotten that it was ever indexed.
 */
data class Bookmark(
    val id: UUID,
    val profileId: UUID,
    val url: String,
    val title: String = "",
    /** The page's own description, or the first paragraph. */
    val excerpt: String = "",
    val siteName: String = "",
    val imageUrl: String? = null,
    /** Name of the Markdown file inside the profile's bookmarks folder. */
    val fileName: String = "",
    /** BCP-47 tag as the page declared it, e.g. `ru`, `en`. */
    val language: String = "",
    /** Length of the readable text — a hint of how much was saved. */
    val characterCount: Int = 0,
    val createdAt: Instant,
    /** When the chunks were embedded. Always null here: this platform does not embed yet. */
    val indexedAt: Instant? = null,
    val embeddingModel: String = "",
    val indexError: String? = null,
    val refreshedAt: Instant? = null,
    /** SHA-256 of the readable text, so a refresh that finds the same page changes nothing. */
    val contentHash: String = "",
    val refreshError: String? = null,
) {
    val lastReadAt: Instant get() = refreshedAt ?: createdAt

    val displayTitle: String get() = title.ifEmpty { url }

    val displayDetail: String
        get() = siteName.ifEmpty { runCatching { java.net.URI(url).host }.getOrNull() ?: url }
}

/**
 * One passage of a bookmark's text, in reading order. Chunk 0 is the title and excerpt, so a search
 * for what a page is about finds it even when the body is long.
 */
data class BookmarkChunk(
    val id: UUID,
    val bookmarkId: UUID,
    val ord: Int,
    val text: String,
)

/** Which bookmarks a search sees: the current profile's, or everyone's. */
enum class BookmarkScope(val id: String) {
    PROFILE("profile"),
    ALL("all"),
    ;

    companion object {
        fun from(id: String?): BookmarkScope = entries.firstOrNull { it.id == id } ?: PROFILE
    }
}

/** A search result: the bookmark, how well it matched, and the passage that matched. */
data class BookmarkHit(
    val bookmark: Bookmark,
    /** 0…1, higher is better. Text matches are fixed, as on the Mac. */
    val score: Double,
    val snippet: String,
)

/**
 * How a saved page becomes a file, a name and a set of passages.
 *
 * All of it is the Mac's arithmetic, because all of it ends up in files and rows that both read.
 *
 * ## One divergence, bounded and named
 *
 * The lengths below are counted in UTF-16 units here and in grapheme clusters on the Mac. For text
 * without astral characters or combining marks the two agree exactly; for text with them a chunk
 * boundary can land a character or two apart. That costs nothing while this platform does not embed
 * — the passages are still the same passages — and it becomes a real question the day it does, since
 * two devices chunking differently produce two different vectors for the same page.
 */
object BookmarkText {

    /** Passages are cut around this many characters; a paragraph longer than [MAX_CHUNK] is split. */
    const val CHUNK_TARGET = 900
    const val MAX_CHUNK = 1400
    const val MAX_CHUNKS = 120

    /** Title and excerpt first, then the text in paragraph-sized passages. */
    fun chunks(title: String, excerpt: String, text: String): List<String> {
        val result = mutableListOf<String>()
        val head = listOf(title, excerpt).filter { it.isNotEmpty() }.joinToString("\n")
        if (head.isNotEmpty()) result.add(head)

        var current = StringBuilder()
        fun flush() {
            val trimmed = current.toString().trim()
            if (trimmed.isNotEmpty()) result.add(trimmed)
            current = StringBuilder()
        }

        for (raw in text.split("\n\n")) {
            val paragraph = raw.trim()
            if (paragraph.isEmpty()) continue
            if (current.length + paragraph.length > CHUNK_TARGET && current.isNotEmpty()) flush()
            if (paragraph.length > MAX_CHUNK) {
                flush()
                result.addAll(split(paragraph, MAX_CHUNK))
            } else {
                if (current.isNotEmpty()) current.append("\n\n")
                current.append(paragraph)
            }
            if (result.size >= MAX_CHUNKS) break
        }
        flush()
        return result.take(MAX_CHUNKS)
    }

    /** Cuts at sentence ends where it can, hard where it must. */
    private fun split(text: String, limit: Int): List<String> {
        val pieces = mutableListOf<String>()
        var rest = text
        while (rest.length > limit) {
            val window = rest.substring(0, limit)
            val lastStop = window.indexOfLast { it in ".!?\n" }
            val cut = if (lastStop >= 0) lastStop + 1 else window.length
            val piece = rest.substring(0, cut).trim()
            if (piece.isNotEmpty()) pieces.add(piece)
            rest = rest.substring(cut)
        }
        val tail = rest.trim()
        if (tail.isNotEmpty()) pieces.add(tail)
        return pieces
    }

    /**
     * `<title slug>-<first 8 of the id>.md`, ASCII-folded so it is the same on any file system.
     *
     * Folded rather than transliterated: a Cyrillic title keeps its letters, which is what the Mac
     * does too — the folding is about accents and case, not alphabet.
     */
    fun fileName(title: String, id: UUID): String {
        val slug = buildString {
            for (character in title.lowercase()) {
                if (character.isLetterOrDigit()) append(character)
                else if (!endsWith("-")) append('-')
            }
        }.trim('-').let { if (it.length > 60) it.take(60).trim('-') else it }

        val short = id.toString().take(8).lowercase()
        return (if (slug.isEmpty()) short else "$slug-$short") + ".md"
    }

    fun hash(text: String): String =
        MessageDigest.getInstance("SHA-256")
            .digest(text.toByteArray())
            .joinToString("") { "%02x".format(it) }

    /**
     * Markdown with YAML front matter — readable in any editor, and enough to rebuild the row.
     *
     * The front matter is what a person opening the folder sees first, so its order is the Mac's and
     * not alphabetical.
     */
    fun document(
        bookmark: Bookmark,
        byline: String,
        profileName: String,
        markdown: String,
    ): String {
        fun quoted(value: String) =
            "\"" + value.replace("\\", "\\\\").replace("\"", "\\\"") + "\""

        val lines = mutableListOf(
            "---",
            "title: ${quoted(bookmark.title)}",
            "url: ${bookmark.url}",
            "site: ${quoted(bookmark.siteName)}",
        )
        if (byline.isNotEmpty()) lines.add("author: ${quoted(byline)}")
        bookmark.imageUrl?.let { lines.add("image: $it") }
        if (bookmark.language.isNotEmpty()) lines.add("language: ${bookmark.language}")
        lines.add("profile: ${quoted(profileName)}")
        lines.add("saved: ${ISO_8601.format(bookmark.createdAt)}")
        lines.add("id: ${bookmark.id.toString().uppercase()}")
        lines.add("---")
        lines.add("")
        if (bookmark.title.isNotEmpty() && !markdown.startsWith("# ")) {
            lines.add("# ${bookmark.title}")
            lines.add("")
        }
        lines.add(markdown)
        lines.add("")
        return lines.joinToString("\n")
    }

    /** `ISO8601DateFormatter()`'s default: seconds, `Z`, no fraction. */
    private val ISO_8601: DateTimeFormatter =
        DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'")
            .withZone(java.time.ZoneOffset.UTC)
}
